test/widgets/action_sheet_test.dart

```diff
@@ -99,6 +103,302 @@ void main() {
+
+      await store.addMessage(message);
+    }
+
+    testWidgets('show from inbox', (tester) async {
+      await prepare();
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: const InboxPage()));
+      await tester.pump();
+
+      await tester.longPress(find.text(topic));
+      // sheet appears onscreen; default duration of bottom-sheet enter animation
+      await tester.pump(const Duration(milliseconds: 250));
+      check(find.byType(BottomSheet)).findsOne();
+    });
+
+    testWidgets('show from app bar', (tester) async {
+      await prepare();
+      connection.prepare(json: eg.newestGetMessagesResult(
+        foundOldest: true, messages: [message]).toJson());
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: MessageListPage(
+          initNarrow: TopicNarrow(channel.streamId, topic))));
+      // global store, per-account store, and message list get loaded
+      await tester.pumpAndSettle();
+
+      await tester.longPress(find.byType(ZulipAppBar));
+      // sheet appears onscreen; default duration of bottom-sheet enter animation
+      await tester.pump(const Duration(milliseconds: 250));
+      check(find.byType(BottomSheet)).findsOne();
+    });
+
+    testWidgets('show from recipient header', (tester) async {
+      await prepare();
+      connection.prepare(json: eg.newestGetMessagesResult(
+        foundOldest: true, messages: [message]).toJson());
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: const MessageListPage(initNarrow: CombinedFeedNarrow())));
+      // global store, per-account store, and message list get loaded
+      await tester.pumpAndSettle();
+
+      await tester.longPress(find.descendant(
+        of: find.byType(RecipientHeader), matching: find.text(topic)));
+      // sheet appears onscreen; default duration of bottom-sheet enter animation
+      await tester.pump(const Duration(milliseconds: 250));
+      check(find.byType(BottomSheet)).findsOne();
+    });
+  });
+
+  group('UserTopicUpdateButton', () {
+    late ZulipStream channel;
+    late String topic;
+
+    Finder findButton({
+      UserTopicVisibilityPolicy? from,
+      required UserTopicVisibilityPolicy to,
+      required String label,
+    }) {
+      return find.descendant(of: find.byWidgetPredicate((widget) {
+        return widget is UserTopicUpdateButton
+          && widget.currentVisibilityPolicy == from
+          && widget.newVisibilityPolicy == to;
+      }), matching: find.text(label));
+    }
+
+    final mute =                 findButton(to:    UserTopicVisibilityPolicy.muted,
+                                            label: 'Mute topic');
+    final unmute =               findButton(from:  UserTopicVisibilityPolicy.muted,
+                                            to:    UserTopicVisibilityPolicy.none,
+                                            label: 'Unmute topic');
```

Comment:
Hmm, this is very specific about the details of how the implementation is arranged.

From the user's perspective it's just a button labeled "Unmute topic", right? And they tap it, and it should produce the right request that unmutes the topic.

So… (next comment below)